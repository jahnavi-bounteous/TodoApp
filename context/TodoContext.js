import React, { createContext, useReducer, useContext, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';

export const ADD_TODO = 'ADD_TODO';
export const DELETE_TODO = 'DELETE_TODO';
export const EDIT_TODO = 'EDIT_TODO';
export const SET_TODOS = 'SET_TODOS';
export const TOGGLE_IMPT = 'TOGGLE_IMP';

const TodoContext = createContext();

export const useTodoContext = () => {
  return useContext(TodoContext);
};

const todoReducer = (state, action) => {
  switch (action.type) {
    case SET_TODOS:
      return action.payload;
    case ADD_TODO:
      return [...state, action.payload];
    case DELETE_TODO:
      return state.filter((todo) => todo.id !== action.payload);
    case EDIT_TODO:
      return state.map((todo) =>
        todo.id === action.payload.id ? action.payload : todo
      );
      case TOGGLE_IMPT:
        return state.map((todo) =>
        todo.id === action.payload ? { ...todo, isImportant: !todo.isImportant } : todo
      );
    default:
      return state;
  }
};

export const TodoProvider = ({ children }) => {
  const [todos, dispatch] = useReducer(todoReducer, []);

  useEffect(() => {
    const loadTodos = async () => {
      try {
        const storedTodos = await AsyncStorage.getItem('todos');
        if (storedTodos !== null) {
          dispatch({ type: SET_TODOS, payload: JSON.parse(storedTodos) });
        }
      } catch (error) {
        console.error('Error loading todos from AsyncStorage:', error);
      }
    };

    loadTodos();
  }, []);

  useEffect(() => {
    const saveTodos = async () => {
      try {
        await AsyncStorage.setItem('todos', JSON.stringify(todos));
      } catch (error) {
        console.error('Error saving todos to AsyncStorage:', error);
      }
    };

    saveTodos();
  }, [todos]);

  return (
    <TodoContext.Provider value={{ todos, dispatch }}>
      {children}
    </TodoContext.Provider>
  );
};