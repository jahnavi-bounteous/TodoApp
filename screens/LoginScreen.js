import React, { useState } from "react";
import { View, Text, StyleSheet, TextInput, TouchableOpacity, Alert } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { MaterialIcons } from '@expo/vector-icons';
import { useTodoContext, SET_TODOS } from '../context/TodoContext'; // Import SET_TODOS and LOGIN_USER

const TodoLoginScreen = ({ navigation}) => {
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [isPasswordVisible, setIsPasswordVisible] = useState(false);
    const { dispatch, loginUser } = useTodoContext(); // Get dispatch function from context

    const handleLoginPress = async () => {
        try {
            const userDataKeys = await AsyncStorage.getAllKeys();
            const userDataEntries = await AsyncStorage.multiGet(userDataKeys.filter(key => key.startsWith('userData_')));
            const users = userDataEntries.map(([_, data]) => JSON.parse(data));
            
            const user = users.find(user => user.username === username && user.password === password);
            if (user) {
                await AsyncStorage.setItem('loggedInUser', JSON.stringify({ username: user.username }));
                loginUser(user.username); // Dispatch LOGIN_USER action
                // Retrieve todos data for the logged-in user
                const userTodosKey = `todos_${user.username}`;
                const todosData = await AsyncStorage.getItem(userTodosKey);
                if (todosData) {
                    const todos = JSON.parse(todosData);
                    console.log(`Todos for user: ${user.username}`, todos);
    
                    // Update todos with user property
                    const todosWithUser = todos.map(todo => ({ ...todo, user: user.username }));
                    dispatch({ type: SET_TODOS, payload: todosWithUser });
                } else {
                    console.log(`No todos found for user: ${user.username}`);
                }
    
                navigation.navigate('Todo');
            } else {
                Alert.alert('Error', 'Invalid username or password. Please try again.');
            }
        } catch (error) {
            console.error('Error retrieving user data from AsyncStorage:', error);
            Alert.alert('Error', 'An error occurred while logging in. Please try again.');
        }
    };   

    return (
        <View style={styles.container}>
            <TextInput
                style={styles.input}
                placeholder="Username"
                value={username}
                onChangeText={(text) => setUsername(text)}
                testID="usernameInput"
            />
            <View style={[styles.passwordContainer, styles.input]}>
                <TextInput
                    style={styles.passwordInput}
                    placeholder="Password"
                    secureTextEntry={!isPasswordVisible}
                    value={password}
                    onChangeText={(text) => setPassword(text)}
                    testID="passwordInput"
                />
                <TouchableOpacity
                    onPress={() => setIsPasswordVisible(!isPasswordVisible)}
                    testID="passwordVisibilityButton"
                >
                    <MaterialIcons name={isPasswordVisible ? 'visibility' : 'visibility-off'} size={24} color="black" />
                </TouchableOpacity>
            </View>
            <TouchableOpacity
                style={[styles.textButton, { backgroundColor: "#1e90ff" }]}
                onPress={handleLoginPress}
                testID="loginButton" 
            >
                <Text style={styles.buttonText}>
                    Login
                </Text>
            </TouchableOpacity>
            <View style={{ flexDirection: 'row' }}>
                <Text>
                    Doesn't have an account ?
                </Text>
                <Text> </Text>
                <TouchableOpacity onPress={() => navigation.navigate('Register')}>
                    <Text style={{ color: 'blue', textDecorationLine: 'underline' }}>
                        Register
                    </Text>
                </TouchableOpacity>
            </View>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    input: {
        width: '80%',
        borderColor: '#1e90ff',
        borderWidth: 1,
        borderRadius: 10,
        fontSize: 16,
        padding: 16,
        marginBottom: 10
    },
    passwordContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
    },
    passwordInput: {
        flex: 1,
        fontSize: 16,
        padding: 0,
    },
    textButton: {
        width: '80%',
        backgroundColor: "#1e90ff",
        borderRadius: 6,
        paddingVertical: 12,
        marginVertical: 10,
        alignItems: "center",
        shadowColor: "#000",
        shadowOffset: { width: 0, height: 2 },
        shadowOpacity: 0.8,
        shadowRadius: 3,
    },
    buttonText: {
        color: '#fff',
        fontSize: 16,
    }
});

export default TodoLoginScreen;

