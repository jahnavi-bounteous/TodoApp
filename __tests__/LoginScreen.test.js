import React from 'react';
import { Text, TouchableOpacity, TextInput } from 'react-native';
import renderer from 'react-test-renderer';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { TodoProvider, useTodoContext } from '../context/TodoContext'; // Import TodoProvider and useTodoContext
import LoginScreen from '../screens/LoginScreen';

// Mock AsyncStorage functions
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
  getAllKeys: jest.fn(),
  multiGet: jest.fn(),
}));

// Mock navigation prop
const mockNavigation = {
  navigate: jest.fn()
};

// Mock useTodoContext hook
jest.mock('../context/TodoContext', () => ({
  ...jest.requireActual('../context/TodoContext'),
  useTodoContext: jest.fn(),
}));

describe('<LoginScreen />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly', () => {
    useTodoContext.mockReturnValue({ dispatch: jest.fn(), loginUser: jest.fn() });
    const tree = renderer.create(
      <TodoProvider>
        <LoginScreen navigation={mockNavigation} />
      </TodoProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('logs in successfully with correct credentials', async () => {
    AsyncStorage.getAllKeys.mockResolvedValueOnce(['userData_test']);
    AsyncStorage.multiGet.mockResolvedValueOnce([['userData_test', '{"username":"test","password":"test"}']]);
    useTodoContext.mockReturnValue({ dispatch: jest.fn(), loginUser: jest.fn() });

    const component = renderer.create(
      <TodoProvider>
        <LoginScreen navigation={mockNavigation} />
      </TodoProvider>
    );
    const instance = component.root;

    const usernameInput = instance.findByProps({ testID: 'usernameInput' });
    const passwordInput = instance.findByProps({ testID: 'passwordInput' });
    const loginButton = instance.findByProps({ testID: 'loginButton' });

    usernameInput.props.onChangeText('test');
    passwordInput.props.onChangeText('test');
    await loginButton.props.onPress();

    expect(AsyncStorage.setItem).toHaveBeenCalledWith('loggedInUser', JSON.stringify({ username: 'test' }));
    expect(mockNavigation.navigate).toHaveBeenCalledWith('Todo');
  });

  test('displays an error message with incorrect credentials', async () => {
    AsyncStorage.getAllKeys.mockResolvedValueOnce(['userData_test']);
    AsyncStorage.multiGet.mockResolvedValueOnce([['userData_test', '{"username":"test","password":"test"}']]);
    useTodoContext.mockReturnValue({ dispatch: jest.fn(), loginUser: jest.fn() });

    const component = renderer.create(
      <TodoProvider>
        <LoginScreen navigation={mockNavigation} />
      </TodoProvider>
    );
    const instance = component.root;

    const usernameInput = instance.findByProps({ testID: 'usernameInput' });
    const passwordInput = instance.findByProps({ testID: 'passwordInput' });
    const loginButton = instance.findByProps({ testID: 'loginButton' });

    usernameInput.props.onChangeText('wrong_user');
    passwordInput.props.onChangeText('wrong_password');
    await loginButton.props.onPress();

    expect(AsyncStorage.setItem).not.toHaveBeenCalled();
    expect(mockNavigation.navigate).not.toHaveBeenCalled();
  });

  test('displays an error message when AsyncStorage throws an error', async () => {
    AsyncStorage.getAllKeys.mockRejectedValueOnce('AsyncStorage error');
    useTodoContext.mockReturnValue({ dispatch: jest.fn(), loginUser: jest.fn() });

    const component = renderer.create(
      <TodoProvider>
        <LoginScreen navigation={mockNavigation} />
      </TodoProvider>
    );
    const instance = component.root;

    const usernameInput = instance.findByProps({ testID: 'usernameInput' });
    const passwordInput = instance.findByProps({ testID: 'passwordInput' });
    const loginButton = instance.findByProps({ testID: 'loginButton' });

    usernameInput.props.onChangeText('test');
    passwordInput.props.onChangeText('test');
    await loginButton.props.onPress();

    expect(AsyncStorage.setItem).not.toHaveBeenCalled();
    expect(mockNavigation.navigate).not.toHaveBeenCalled();
  });
});
