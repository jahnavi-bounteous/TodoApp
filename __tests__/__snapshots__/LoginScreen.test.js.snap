// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Username"
    style={
      {
        "borderColor": "#1e90ff",
        "borderRadius": 10,
        "borderWidth": 1,
        "fontSize": 16,
        "marginBottom": 10,
        "padding": 16,
        "width": "80%",
      }
    }
    testID="usernameInput"
    value=""
  />
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        {
          "borderColor": "#1e90ff",
          "borderRadius": 10,
          "borderWidth": 1,
          "fontSize": 16,
          "marginBottom": 10,
          "padding": 16,
          "width": "80%",
        },
      ]
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Password"
      secureTextEntry={true}
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "padding": 0,
        }
      }
      testID="passwordInput"
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="passwordVisibilityButton"
    >
      <Text />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#1e90ff",
        "borderRadius": 6,
        "marginVertical": 10,
        "opacity": 1,
        "paddingVertical": 12,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 3,
        "width": "80%",
      }
    }
    testID="loginButton"
  >
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 16,
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <Text>
      Doesn't have an account ?
    </Text>
    <Text>
       
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "blue",
            "textDecorationLine": "underline",
          }
        }
      >
        Register
      </Text>
    </View>
  </View>
</View>
`;
